linera-explorer/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+        .push_state_with_url(&page_js, &new_path, Some(&new_path))
+        .expect("push_state failed");
+}
+
+#[wasm_bindgen]
+pub async fn route(app: JsValue, path: JsValue, args: JsValue) {
+    let path = path.as_string();
+    let args = from_value::<Vec<(String, String)>>(args).unwrap_or(vec![]);
+    let msg = format!(
+        "route: {} {:?}",
+        path.clone().unwrap_or("none".to_string()),
+        args
+    );
+    log_str(&msg);
+    let data = from_value::<Data>(app.clone()).expect("cannot parse vue data");
+    let chain_id = args
+        .iter()
+        .find_map(|(k, v)| {
+            if k == "chain" {
+                Some(ChainId::from_str(v).expect("wrong chain id"))
+            } else {
+                None
+            }
+        })
+        .unwrap_or(data.chain);
+    route_aux(&app, &data, &path, &chain_id, &args).await
+}
+
+#[wasm_bindgen]
+pub fn short(s: String) -> String {
+    let n = s.len();
+    format!("{}..{}", &s[..4], &s[n - 4..])
+}
+
+#[wasm_bindgen]
+pub fn short_app(s: String) -> String {
+    format!("{}..{}..{}..", &s[..4], &s[64..68], &s[152..156])
+}
+
+fn set_onpopstate(app: JsValue) {
+    let callback = Closure::<dyn FnMut(JsValue)>::new(move |v: JsValue| {
+        setf(&app, "page", &getf(&v, "state"));
+    });
+    web_sys::window()
+        .expect("window object not found")
+        .set_onpopstate(Some(callback.as_ref().unchecked_ref()));
+    callback.forget()
+}
+
+/// Initialize pages and subscribe to notifications
+#[wasm_bindgen]
+pub async fn init(app: JsValue, uri: String) {
+    console_error_panic_hook::set_once();
+    set_onpopstate(app.clone());
+    let data = from_value::<Data>(app.clone()).expect("cannot parse vue data");
+    let address = node_service_address(&data.config, false);
+    let default_chain = chains(&app, &address).await;
+    match default_chain {
+        Err(e) => {
+            route_aux(
+                &app,
+                &data,
+                &Some("error".to_string()),
+                &data.chain,
+                &[("msg".to_string(), e.to_string())],
+            )
+            .await;
+        }
+        Ok(default_chain) => {
+            let uri = Url::parse(&uri).expect("failed to parse url");
+            let pathname = uri.path();
+            let mut queries = uri.query_pairs();
+            let chain_id = queries
+                .find_map(|(k, v)| {
+                    if k == "chain" {
+                        Some(ChainId::from_str(&v).expect("cannot build chainId"))
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or(default_chain);
+            let (path, args) = match pathname {
+                "/blocks" => (Some("blocks".to_string()), vec![]),
+                "/applications" => (Some("applications".to_string()), vec![]),
+                pathname => {
+                    let path = std::path::Path::new(pathname);
```

Comment:
Why `std::path::Path`? `str` also has a `starts_with` method. Or does this normalize slashes that `uri.path()` might have duplicated or omitted?